ccadb/ccadb.go

```diff
@@ -0,0 +1,289 @@
+	}, nil
+}
+
+func (c *Checker) Check(ctx context.Context) error {
+	crlURLs, err := c.getCRLURLs(ctx, c.allCertificatesCSVURL, c.caOwner)
+	if err != nil {
+		return err
+	}
+
+	var crls, entries, bytes int
+
+	serials := make(map[string]*x509.RevocationList)
+
+	var errs []error
+	for skid, urls := range crlURLs {
+		for _, url := range urls {
+			crls++
+			issuer := c.issuers[skid]
+			if issuer == nil {
+				return fmt.Errorf("no issuer found for skid %x", skid)
+			}
+			crl, err := checkCRL(ctx, url, issuer, c.crlAgeLimit)
+			if err != nil {
+				errs = append(errs, fmt.Errorf("fetching %s: %s", url, err))
+				continue
+			}
+
+			// Check for duplicates across different CRLs (or within a CRL).
+			// Cap any given CRL at 1M entries to limit memory use.
+			for i, entry := range crl.RevokedCertificateEntries {
+				if i > 1_000_000 {
+					break
+				}
+				serialByteString := string(entry.SerialNumber.Bytes())
+				if otherCRL, ok := serials[serialByteString]; ok {
+					otherCRLURL, err := idp.Get(otherCRL)
+					if err != nil {
+						return err
+					}
+					errs = append(errs, fmt.Errorf("serial %x seen on multiple CRLs: %s and %s", entry.SerialNumber, otherCRLURL, url))
+				}
+				serials[serialByteString] = crl
+			}
+
+			age := time.Since(crl.ThisUpdate).Round(time.Minute)
+			nextUpdate := time.Until(crl.NextUpdate).Round(time.Hour)
+			entries += len(crl.RevokedCertificateEntries)
+			bytes += len(crl.Raw)
+			log.Printf("crl %q: %d entries, %d bytes, age %gm, nextUpdate %gh", url, len(crl.RevokedCertificateEntries), len(crl.Raw), age.Minutes(), nextUpdate.Hours())
+		}
+	}
+
+	log.Printf("%d CRLs had %d entries and %d bytes", crls, entries, bytes)
+	return errors.Join(errs...)
+}
+
+func checkCRL(ctx context.Context, url string, issuer *x509.Certificate, ageLimit time.Duration) (*x509.RevocationList, error) {
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	if err != nil {
+		return nil, err
+	}
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("HTTP status code %d", resp.StatusCode)
+	}
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("reading CRL body: %s", err)
+	}
+	crl, err := x509.ParseRevocationList(body)
+	if err != nil {
+		return nil, err
+	}
+
+	idp, err := idp.Get(crl)
+	if err != nil {
+		return nil, err
+	}
+
+	if idp != url {
+		return nil, fmt.Errorf("CRL fetched from %s had mismatched IDP %s", url, idp)
+	}
+
+	return crl, checker.Validate(crl, issuer, ageLimit)
+}
+
+// returns a map from issuer SKID to list of URLs
+func (c Checker) getCRLURLs(ctx context.Context, csvURL string, owner string) (map[string][]string, error) {
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, csvURL, nil)
+	if err != nil {
+		return nil, err
+	}
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
How do we want to alert on this service failing?

If a single failure run could indicate a problem that we need to attend to immediately, then we should retry with exponential backoff on this call so that transient failures don't wake us up.

Reply:
My expectation is that we'll deploy the lambda (lambda/ccadbchecker) and alert on any error it returns.

Thanks for spotting the need for retries. I had done some of the necessary refactoring in https://github.com/letsencrypt/crl-monitor/pull/89/files#diff-cb4d75d4fe1fdb084eb3e51210725f1a6f801c2b800f996b4197f74f4a105d86 but forgot to apply that to this PR. Done.